services/packagist/packagist-license.service.js

```diff
@@ -59,17 +56,18 @@ module.exports = class PackagistLicense extends BasePackagistService {
   }
 
   transform({ json, user, repo }) {
-    const branch = this.getDefaultBranch(json, user, repo)
-    if (!branch) {
-      throw new NotFound({ prettyMessage: 'default branch not found' })
-    }
-    const { license } = branch
```

Comment:
tbh I'm not sure I know what we were doing here before. Do you know what, if any, the chances are of this API change resulting in us finding and rendering a different license because of the branch change?

Reply:
It shouldn't ever find or render the wrong license because we only care about the most recent version which is always the first item in the `json.packages[packageName]` array.

Reply:
Thanks, though my question isn't about "right" or "wrong". We take backwards compatibility pretty seriously, and one of the things we have to be mindful of is whether a change we make to the Shields code has the potential to modify any of our existing badge users without them changing anything in their corresponding project/repo/package/etc.

It's still not clear to me whether that's a possibility here

Reply:
I guess it is technically possible here but there is only one way and it's likely never going to be an issue. The only way it could be a problem is if the default branch has a different license than the most recent tagged release. For example, if a project had an MIT license before (so in the most recent tagged release) but you changed it to Apache License 2.0 in your default branch, main, master, dev, whatever you set your default branch to, and did not tag a new release. The current Shields, with the old api, would return Apache License 2.0 but this PR would return MIT because it's the license of the most recent tagged release.

Reply:
That's helpful clarity, thank you!

Based on a cursory glance of the docs under https://packagist.org/apidoc#getting-package-data it seems like we could maintain access to the full set of dev/branch data if we fetched `packageName-dev.json` instead of `packageName.json`

Do you see any issues with that?

Reply:
There are multiple different issues tied together here. It is a bit messy and complicated..

- The old packagist `/p` API used to return both tagged releases and dev versions (branches) in the same API response. This was always a somewhat esotric behaviour of packagist in comparison to other package registries.
- The new packagist API separates tagged and dev versions into different API endpoints: https://packagist.org/apidoc#get-package-data So https://packagist.org/p2/symfony/symfony.json now returns tagged releases and https://packagist.org/p2/symfony/symfony~dev.json returns dev releases.
- Two of the packagist badges (`/l` and `/php-v`) have long used the default branch rather than the latest version as the reference point for license/php-version. This is one of those things that mostly goes unnoticed (esp for the license case because projects change their license very infrequently), but it is actually a long-standing bug. See https://github.com/badges/shields/issues/2356 It never got fixed, but I'd acknowledge that the existing behaviour is a bug and the correct behaviour is to default to the **latest stable version** not the default branch if `/php-v` or `/l` is called with no explicit version (i.e: a call like https://img.shields.io/packagist/php-v/doctrine/orm rather than https://img.shields.io/packagist/php-v/doctrine/orm/2.8.4 ). Same as we would do for any other package registry. We shouldn't go out of our way to call the `~dev` endpoint to reproduce the existing behaviour IMO. We should use this opportunity to fix it.
- That said, the code here (just using the first item in the array) does not give us the latest stable version. For example, if I call https://packagist.org/p2/symfony/symfony.json the first version in the array is `v5.3.0-RC1` (a pre-release). If we're going to properly fix this we should parse the array for the first stable version and use that, not just pick the first release in the array. It is less of an issue for license, but more important for the supported PHP versions badge.
- Another related issue/knock-on effect is that there are some badge queries which are currently valid which will become invalid with this approach. For example at the moment https://img.shields.io/packagist/php-v/cakephp/cakephp/dev-master is a valid badge but once we switch from `/p` to `/p2` that will become an error because `dev-master` isn't a tagged version. I think (again for the `/php-v` endpoint), if it called with the `:version` param we could:
    - call `/p2/user/repo.json`
    - look for `version` in the tagged releases
    - if we don't find it, make a second call to the `/p2/user/repo~dev.json` endpoint to look for it 
without adding too much extra complexity to keep that variant working as it does now.
- Also mixed up in this is the fact that the new `/p2` API endpoint doesn't return all of the keys for all of the version objects. This is sufficiently unexpected that I raised an issue about it over on the packagist repo: https://github.com/composer/packagist/issues/1194 It turns out this is intended and also adds a bit of extra complexity for us: The recommendation is we port https://github.com/composer/metadata-minifier/blob/c549d23829536f0d0e984aaabbf02af91f443207/src/MetadataMinifier.php#L22-L46 to javascript to 'decompress' the metadata before we use it. Fortunately this doesn't look to complex. So having said all that, I think the correct approach is:

PHP version support (`/php-v`) badge:
- Call `/p2/:user/:repo.json`
- Decompress the response using a port of https://github.com/composer/metadata-minifier/blob/c549d23829536f0d0e984aaabbf02af91f443207/src/MetadataMinifier.php#L22-L46
- If we are called with no `:version` param, identify the latest stable release. If there is none, fall back to the latest pre-release (first item in the array). Report the PHP version support for that release.
- If we are called with a `:version` param, try to find that release. If it does exist, fine. If not, make a second call to `/p2/:user/:repo~dev.json`, decompress that and see if we can find it there. If we find it in either, report the PHP version support for that release.

License badge (`/l`):
- Call `/p2/:user/:repo.json`
- Decompress the response using a port of https://github.com/composer/metadata-minifier/blob/c549d23829536f0d0e984aaabbf02af91f443207/src/MetadataMinifier.php#L22-L46
- Identify the latest stable release. If there is none, fall back to the latest pre-release. Report the PHP license for that release.

Sorry - this is a bit of a can-of-worms for a first contribution :grimacing:

Reply:
Looks like porting the metadata expander is pretty straightforward so I should be able to knock these changes out when I have some time later tonight or tomorrow 👍

Reply:
> The new packagist API separates tagged and dev versions into different API endpoints:

Ah okay. That's what I get for not actually looking at the responses :laughing: 

> We shouldn't go out of our way to call the ~dev endpoint to reproduce the existing behaviour IMO. We should use this opportunity to fix it.

Makes sense, and agreed :+1: 

> So having said all that, I think the correct approach is: .....

SGTM

Reply:
I'm taking a look at Composer's MetadataMinifier package and I realized that the expand method doesn't actually do what we need. All it's doing is "expanding" the array of releases with data we already have. In other words, it doesn't do anything useful if we want to get the PHP version requirements or license (or anything really) for the latest stable release if it's not the first item in the array.

Take a simplified version of the response from `/p2/symfony/symfony.json`:
```json
"packages": {
  "symfony/symfony": [
    {
      "version": "v5.3.0-RC1",
      "license": [
        "MIT"
      ],
      "require": {
        "php": ">=7.2.5"
      },
    },
    {
      "version": "v5.3.0-BETA4",
    },
    {
      "version": "v5.3.0-BETA3",
    },
    {
      "version": "v5.3.0-BETA2",
    },
    {
      "version": "v5.3.0-BETA1",
    },
    {
      "version": "v5.2.9",
      "require": {
        "php": ">=7.2.5",
      },
    }
  ]
}
```

After "expanding", `license` for `v5.2.9 (latest stable release) will be whatever `license` is for the first item, `v5.3.0-RC1 in this case. In other words, if Symfony were to change its license in a pre-release,  that would be the license used for the badge regardless. The same could go for PHP version requirements though not in the case of Symfony right now.

The only way to ensure we actually have the right data is to use the standard JSON API like `packagist.org/packages/symfony/symfony.json`. This is a more expensive request because the response is massive, but it does respond with all of the data we would need for every release. This poses some other issues as well. As far as I can tell, it still wouldn't return any branches/dev versions like `dev-master` or `dev-main`. If we needed `dev` we could fall back to `/p2/user/repo~dev.json`. The other issue is that it seems the response is in no particular order. It's not a huge deal but determining the most recent stable release could be a little time-consuming since it would have to go through the entire response. 

I'll set some time aside tomorrow to start tackling this.

Reply:
This is not the case. Just trust the code instead of trying to infer things :) If the license changed in a prerelease, then the minified json would look like:

```jsonc
"packages": {
  "symfony/symfony": [
    {
      "version": "v5.3.0-RC1",
      "license": [
        "NEW-5.3-LICENSE" // new license here
      ],
      "require": {
        "php": ">=7.2.5"
      },
    },
    {
      "version": "v5.3.0-BETA4",
    },
    {
      "version": "v5.3.0-BETA3",
    },
    {
      "version": "v5.3.0-BETA2",
    },
    {
      "version": "v5.3.0-BETA1", // new license was added in beta1 let's assume, so it is valid until here
    },
    {
      "version": "v5.2.9",
      "require": {
        "php": ">=7.2.5",
      },
      "license": [
        "MIT" // old license now is set as it changed from the versions above
      ],
    }
  ]
}
```

In both cases, after running MetadataMinifier::expand, you would have `license:["MIT"]` present in v5.2.9.

`/p2/user/repo~dev.json` includes *only* dev releases, so if you're looking for latest stable this is entirely useless. Please just use `/p2/user/repo.json` with a MetadataMinifier::expand port.

Reply:
Also, please everyone take a couple minutes to read through https://packagist.org/apidoc#get-package-data which explains our new v2 api in details. Note the point about caching and `If-Modified-Since`.. I don't know if you can easily integrate this here to save everyone some bandwidth but it would be nice.

Reply:
> This is not the case. Please trust the code instead of trying to infer things :) If the license changed in a prerelease, then the minified json would look like:

@Seldaek so you’re saying that if there is a change between lisences, the minified response would still include the license with the last version to have that license before the change? I guess I just didn’t fully understand what was minified in the response.

Reply:
If you iterate through versions, the license will always be the last license you saw above. If it isn't anymore, then the license is present in the version object. If an older version lacks a license, the license is set to `"__unset"`. But really IMO the best is to run the `expand` logic and *then* use the objects as you would normally, so you don't need to consider any of this.